### PR TITLE
Add compiler CI

### PR DESCRIPTION
**Pull request by JonasWanke**
There are a few Clippy complaints left:

* AST's `CallReceiver::Call` is not yet created while parsing, but is already handled in the HIR, hence I didn't want to remove it. The parser handling can be done separately.
* Some modules (`fuzzer`, `vm`) contain modules with the same name – do we want to give them a different name or ignore this lint?
* Two `to_*` methods are not idiomatic with respect to using a reference or consuming the value.
* and a few unused methods

### CODE REVIEW

File: compiler/src/compiler/string_to_rcst.rs [hunk 1]
```diff
@@ -80,16 +80,12 @@ mod parse {
...
+    static SUPPORTED_WHITESPACE: &str = " \r\n\t";
 
     fn literal<'a>(input: &'a str, literal: &'static str) -> Option<&'a str> {
         log::trace!("literal({input:?}, {literal:?})");
-        if input.starts_with(literal) {
-            Some(&input[literal.len()..])
-        } else {
-            None
-        }
+        input.strip_prefix(literal)
```
Comment (MarcelGarus) [inline compiler/src/compiler/string_to_rcst.rs:88]: Whoa, was that detected by clippy? Impressive!

**Review by MarcelGarus — APPROVED** [review]

**Comment by MarcelGarus** [comment]
About the reported lints:

- Regarding `CallReceiver`: What exactly is the benefit of analyzing a call's receiver directly in the AST over just storing an arbitrary `Ast` expression as the receiver? Like, sure, we'll find out in the AST that the call is invalid if the receiver is an int literal or something like that, but we would also find those cases during constant evaluation. So, in terms of developer experience, it's not really a difference.
  What's more, in the next stage of the compiler pipeline, the AST-to-HIR conversion, we don't actually use the receiver information in a meaningful way: After special-casing the calling of a  `AstIdentifier::("needs")`, we just match over the receiver type, and if it's a `StructAccess`, we call `lower_struct_access`, if it's a `Call`, we call `lower_call`, and if it's an `Identifier`, we essentially have a copied version of the `compile_single`' code for lowering an `Identifier`. We could just as easily treat all receivers the same (just `compile_single` them directly).
- Regarding modules with the same name: I think it makes sense to ignore this lint. For example, I like that modules can have local `utils` modules without polluting some outer scope.
- Regarding the `to_*` methods: I guess it makes sense to rename them then.
- Few unused methods: That's probably okay for now. Some of them, such as `Hir::return_value`, at least feel like they could become useful later on. But in the long term, ofc it makes sense to clean these up too.

**Comment by JonasWanke** [comment]
Regarding the modules with the same name: I think my explanation wasn't clear, so here's a link to the lint: https://rust-lang.github.io/rust-clippy/master/index.html#module_inception. The problem is module `fuzzer` _inside_ module `fuzzer`

File: compiler/src/compiler/cst_to_ast.rs [hunk 2]
```diff
@@ -233,12 +232,12 @@ impl LoweringContext {
...
-                        CallReceiver::StructAccess(self.lower_struct_access(struct_, dot, key)),
-                    ),
+                    CstKind::Identifier(identifier) => {
+                        Some(self.lower_identifier(name.id, identifier.to_string()))
+                    }
+                    CstKind::StructAccess { struct_, dot, key } => {
+                        Some(self.lower_struct_access(name.id.to_owned(), struct_, dot, key))
+                    }
                     _ => None,
                 };
```
Comment (MarcelGarus) [inline compiler/src/compiler/cst_to_ast.rs]: Doesn't the same reasoning that motivated the removal of a `CallReceiver` also apply here? i.e. not special-casing based on the type of the receiver and just doing `let receiver = self.lower_cst(name);` That would also allow for `lower_identifier` and `lower_struct_access` to be inlined.

Sidenote: It probably also makes sense to rename `CstKind::Call`'s `name` to `receiver`.

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

**Comment by MarcelGarus** [comment]
Regarding the `module_inception` lint: That's a fair concern, I agree that having both a `mod.rs` and `fuzzer.rs` in `fuzzer` is kinda weird. We could merge both into the `mod.rs` file, but they _do_ contain different logic (the CLI and exports vs. the actual fuzzing logic).

In the best case, I think we'd rename the `fuzzer.rs` to something else, but I'm having a hard time finding a fitting name. We could try to go for a name that describes the module's role _within_ the fuzzer, i.e. `engine` or something like that. What do you think?

**Review by MarcelGarus — APPROVED** [review]

**Comment by MarcelGarus** [comment]
Just noticed that the term `receiver` probably stems from object-oriented programming languages, where method invocations are implemented using message passing and a better term in Candy would probably be `callee`. But we can change that later.

**Pull request merged**